packages/arb-token-bridge-ui/src/hooks/usePendingTransactions.ts

```diff
@@ -0,0 +1,144 @@
+import { Provider } from '@ethersproject/providers'
+import dayjs from 'dayjs'
+import {
+  fetchDeposits,
+  fetchWithdrawals,
+  OutgoingMessageState
+} from 'token-bridge-sdk'
+import {
+  outgoungStateToString,
+  transformDeposits,
+  transformWithdrawals
+} from '../state/app/utils'
+import useSWR from 'swr'
+import { useNetworksAndSigners } from '../hooks/useNetworksAndSigners'
+import { useAppState } from '../state'
+import { useGateways } from './useGateways'
+import { DepositStatus } from '../state/app/state'
+
+const INITIAL_PAGE_SIZE = 10
+
+export const fetchPendingTransactions = async ({
+  address,
+  l1Provider,
+  l2Provider,
+  gatewayAddresses,
+  pageSize = INITIAL_PAGE_SIZE
+}: {
+  address: string
+  l1Provider: Provider
+  l2Provider: Provider
+  gatewayAddresses: string[]
+  pageNumber?: number
+  pageSize?: number
+  searchString?: string
+}) => {
+  console.log('***** START: fetching pending transactions *****')
+  // fetch the first 100 deposits
+  const deposits = await fetchDeposits({
+    address,
+    l1Provider,
+    l2Provider,
+    pageNumber: 0,
+    pageSize
+  })
+  console.log('***** fetched pending deposits *****')
+
+  console.log('***** fetching pending withdrawals *****')
+  // fetch the first 100 withdrawals
+  const withdrawals = await fetchWithdrawals({
+    address,
+    l1Provider,
+    l2Provider,
+    pageNumber: 0,
+    pageSize,
+    gatewayAddresses
+  })
+  console.log('***** fetched pending withdrawals *****')
+
+  // filter out pending deposits
+  const pendingDepositsMap: { [id: string]: boolean } = {}
+  // get their complete transformed data (so that we get their exact status)
+  const completeDepositData = transformDeposits(deposits)
+  completeDepositData.forEach(completeTxData => {
+    if (
+      completeTxData &&
+      completeTxData.depositStatus !== DepositStatus.L2_SUCCESS
+    ) {
+      pendingDepositsMap[completeTxData.txId] = true
+    }
+  })
+  const pendingDeposits = deposits.filter(tx => pendingDepositsMap[tx.txID])
+
+  // filter out pending withdrawals
+  const pendingWithrawalMap: { [id: string]: boolean } = {}
```

Comment:
If it behaves like a map, and is named like a map, can we use `Map` instead 😁?

Reply:
Or we might only need a set here. And on line 85, we can use `const pendingWithdrawals = withdrawals.filter(tx => pendingWithdrawalSet.has(tx))`

Reply:
Done